File: constrained-device-app/src/main/python/programmingtheiot/cda/system/BaseSystemUtilTask.py
#####
# 
# This class is part of the Programming the Internet of Things project.
# 
# It is provided as a simple shell to guide the student and assist with
# implementation for the Programming the Internet of Things exercises,
# and designed to be modified by the student as needed.
#

from programmingtheiot.data.SensorData import SensorData
import programmingtheiot.common.ConfigConst as ConfigConst
import logging
import random 

class BaseSystemUtilTask():
    """
    Shell representation of class for base system utility task.
    """
    
    def __init__(self, sensorName = ConfigConst.NOT_SET):
        """
        Constructor
        
        @param sensorNAme: default to Not Set
        """
        self.sensorName = sensorName
        self.latestSensorData = None
        self.minVal = 0.0
        self.maxVal = 100.0
    
    def generateTelemetry(self) -> SensorData:
        """
        A method to generate SensorData
        
        @return SensorData
        """
        ###
        # NOTE: Use self._getSystemUtil() to retrieve the value from the sub-class
        # Create a new SensorData instance and 
        sd = SensorData()
        # Set the value of new SensorData to be the return value of self._getSystemUtil()
        sd.setValue(self._getSystemUtil())
        # Set a new SensorData as the latestSensorData reference
        self.latestSensorData = sd
        # Return the latestSensorData
        return self.latestSensorData
        
    def getTelemetryValue(self) -> float:
        """
        A method to get a telemetry value
        
        @return float
        """
        # If latestSensorData == None, call generateTelemetry()
        if self.latestSensorData == None: self.generateTelemetry()
        # Get the telemetry value from latestSensorData
        val = self.latestSensorData.getValue()
        logging.info("{}: Telemetry value = {}".format(self.__class__.__name__, str(val)))
        return val
    
    def _getSystemUtil(self) -> float:
        """
        Template method implemented by sub-class.
        
        Retrieve the system utilization value as a float.
        
        @return float
        """
        pass
        

File: gateway-device-app/src/main/java/programmingtheiot/gda/system/BaseSystemUtilTask.java
/**
 * This class is part of the Programming the Internet of Things project.
 * 
 * It is provided as a simple shell to guide the student and assist with
 * implementation for the Programming the Internet of Things exercises,
 * and designed to be modified by the student as needed.
 */ 

package programmingtheiot.gda.system;

import java.util.logging.Logger;

import programmingtheiot.data.SensorData;

/**
 *
 */
public abstract class BaseSystemUtilTask
{
	// static
	
	private static final Logger _Logger =
		Logger.getLogger(BaseSystemUtilTask.class.getName());
	
	// private
	private SensorData latestSensorData = null;
	
	// constructors
	
	public BaseSystemUtilTask()
	{
		super();
	}
	
	
	// public methods
	
	public SensorData generateTelemetry()
	{
		// Create a new SensorData instance and 
		SensorData sd = new SensorData();
		// Set the value of latestSensorData to be the return value of getSystemUtil()
		sd.setValue(getSystemUtil());
		// Set new SensorData as the latestSensorData reference
		this.latestSensorData = sd;
		// Return the latestSensorData
		return this.latestSensorData;
	}
	
	public float getTelemetryValue()
	{
		if (this.latestSensorData == null) this.generateTelemetry();
		float val = this.generateTelemetry().getValue();
		_Logger.info("Get telemetry value = " + val);
		return val;
	}
	
	
	// protected methods
	
	/**
	 * Template method definition. Sub-class will implement this to retrieve
	 * the system utilization measure.
	 * 
	 * @return float
	 */
	protected abstract float getSystemUtil();
	
}
